firedrake/solving_utils.py

```diff
@@ -21,6 +23,41 @@ def _make_reasons(reasons):
 SNESReasons = _make_reasons(PETSc.SNES.ConvergedReason())
 
 
+if get_config()["options"]["petsc_int_type"] == "int32":
+    DEFAULT_KSP_PARAMETERS = {"mat_type": "aij",
+                              "ksp_type": "preonly",
+                              "pc_type": "lu",
+                              "pc_factor_mat_solver_type": "mumps"}
+else:
+    # FIXME: which factorisation package to use in int64?
+    DEFAULT_KSP_PARAMETERS = {}
+
+
+def set_defaults(solver_parameters, arguments, *defaults):
+    """Set defaults for solver parameters.
+
+    :arg solver_parameters: dict of user solver parameters to override/extend defaults
+    :arg arguments: arguments for the bilinear form (need to know if we have a Real block).
+    :arg defaults: Any default sets of parameters to supply."""
+    parameters = dict(chain.from_iterable(d.items() for d in defaults))
+    if any(isinstance(V, fsi.RealFunctionSpace)
+           for V in chain.from_iterable(a.function_space() for a in arguments)):
+        # There's a Real block, override with defaults with nest + jacobi.
+        parameters.update({"pc_type": "jacobi", "mat_type": "nest"})
+    solver_parameters = solver_parameters or {}
+    mat_type = solver_parameters.get("mat_type")
+    pmat_type = solver_parameters.get("pmat_type")
+    # Allow anything, interpret "matfree" as matrix_free.
+    matfree = mat_type == "matfree"
+    pmatfree = pmat_type == "matfree"
+    # No preconditioner by default for matrix-free
+    if matfree or pmatfree:
+        parameters.update({"pc_type": "none"})
```

Comment:
Should we automatically set `ksp_type: "fgmres"` and `"ksp_monitor_true_residual": None`, too? It will default to `preonly/none` if the user just passes `{"mat_type": "matfree"}`, which probably is undesirable.

Reply:
True. I realised when doing this in general that setting up defaults for all the different cases is somewhat fiddly. I'm inclined to do:

1. No options, no real blocks -> LU + preonly
2. No options, real blocks -> either Jacobi or auto-generate fieldsplit options + LU
3. Anything else: no defaults, user knows what they're doing.

Reply:
I agree (with choice fieldsplit + LU for point 2).

Reply:
Can we detect the fieldsplit reliably enough? E.g. how do we determine which block to eliminate?

Reply:
If you have a mixed system then even in 2D LU can get big fast and/or defeat the MUMPS memory management heuristics. Is this really a robust option?

Reply:
I'm happy to add superlu_dist to the default set of installed packages (it knows about malloc!). In terms of "big", 2D problems with O(1-10e6) dofs are quite happily solved in reasonable time using MUMPs on a workstation.

In terms of defaults, note that in all our tutorials we explicitly say "use LU until you have your numerics/model right and then think about preconditioning". But then we don't default to that option, so you have to learn a load of magic incantations first.

Reply:
I've changed to (hopefully) do the 3-part proposal above (with fieldsplit + Schur elimination for problems with Real blocks).